mmengine/utils/progressbar_rich.py

```diff
@@ -0,0 +1,365 @@
+        assert total > 0
+        self.colors.append(color)
+        self.tasks.append(
+            self.bar.add_task(f'[{color}]{description}', total=total))
+
+    def add_single_task(self, total, color='blue', description='Process...'):
+        self.colors.append(color)
+        if total > 0:
+            self.tasks.append(
+                self.bar.add_task(f'[{color}]{description}', total=total))
+            self.infinite = False
+        else:
+            self.write('completed: 0, elapsed: 0s', self.colors[0])
+            self.timer = Timer()
+            self.infinite = True
+            self.completed = 0
+            self.tasks.append(0)
+
+    def is_task_finish(self, task_id):
+        completed = self.bar.tasks[task_id].completed
+        total = self.bar.tasks[task_id].total
+        return completed == total
+
+    def finished(self):
+        return self.bar.finished
+
+    def update(self, task_id=0, advance=1):
+        assert advance > 0
+
+        if len(self.tasks) == 1:
+            assert task_id == 0
+            if not self.infinite:
+                self.bar.advance(self.tasks[task_id], advance=advance)
+
+                if self.is_task_finish(task_id):
+                    self.bar.stop()
+            else:
+                self.completed += advance
+                elapsed = self.timer.since_start()
+                if elapsed > 0:
+                    fps = self.completed / elapsed
+                else:
+                    fps = float('inf')
+                self.write(
+                    f'completed: {self.completed},'
+                    f' elapsed: {int(elapsed + 0.5)}s,'
+                    f' {fps:.1f} tasks/s', self.colors[task_id])
+        else:
+            assert task_id >= 0 and task_id < len(self.tasks)
+            self.bar.update(self.tasks[task_id], advance=advance)
+            time.sleep(0.01)
+
+            if self.bar.finished:
+                self.bar.stop()
+
+    def reset(self):
+        for i in range(len(self.tasks)):
+            self.bar.remove_task(self.tasks[i])
+        self.tasks.clear()
+        self.colors.clear()
+
+
+def track_single_progress(func,
+                          tasks,
+                          description='Process...',
+                          color='blue',
+                          **kwargs):
+    """Track the progress of tasks execution with a progress bar.
+
+    Tasks are done with a simple for-loop.
+
+    Args:
+        func (callable): The function to be applied to each task.
+        tasks (list or tuple[Iterable, int]): A list of tasks or
+            (tasks, total num).
+        description (str): The description of progress bar.
+        color (str): The color of progress bar.
+
+    Returns:
+        list: The task results.
+    """
+    if isinstance(tasks, tuple):
+        assert len(tasks) == 2
+        assert isinstance(tasks[0], Iterable)
+        assert isinstance(tasks[1], int)
```

Comment:
How about setting `task_num` and `task` as two arguments? The default value for `task_num` is `None`, which could be inferred by `len(task_num)`. If `tasks` is an iterator that does not implement `__len__` method,  users can pass the `num_tasks` arguments manually or just pass `total=None` to the progress bar (only print the update message without the rich bar,).